compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2605,6 +2606,27 @@ impl<'a> Parser<'a> {
         if let ExprKind::Let(_, _, _, Recovered::No) = cond.kind {
             // Remove the last feature gating of a `let` expression since it's stable.
             self.psess.gated_spans.ungate_last(sym::let_chains, cond.span);
+        } else {
+            fn ungate_let_exprs(this: &mut Parser<'_>, expr: &Expr) {
+                if !expr.span.at_least_rust_2024() {
+                    return;
+                }
```

Comment:
I'd like to see some tests on this behavior where a nested expression does not get ungated, perhaps from a macro in edition 2024 where arguments are in edition 2021:

```rust
macro_rules! test {
    ($($tt:tt)*) => {
        if $($tt)* {
            println!("meow")
        }
    };
}

fn main() {
    test!(let None = None::<()> && let None = None::<()>);
}
```

Reply:
Added such a test going through various possible cases.